xray/grpc.go

```diff
@@ -0,0 +1,143 @@
+package xray
+
+import (
+	"context"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"net/http"
+	"net/http/httptrace"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-xray-sdk-go/header"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// UnaryClientInterceptor provides gRPC unary client interceptor.
+func UnaryClientInterceptor(host string) grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		return Capture(ctx, host, func(ctx context.Context) error {
+			seg := GetSegment(ctx)
+			if seg == nil {
+				return errors.New("failed to record gRPC transaction: segment cannot be found")
+			}
+
+			ct, e := NewClientTrace(ctx)
+			if e != nil {
+				return e
+			}
+			md := metadata.Pairs(TraceIDHeaderKey, seg.DownstreamHeader().String())
+			ctx2 := metadata.NewOutgoingContext(httptrace.WithClientTrace(ctx, ct.httpTrace), md)
+
+			seg.Lock()
+			seg.Namespace = "remote"
+			seg.GetHTTP().GetRequest().URL = "grpc://" + host + method
+			seg.GetHTTP().GetRequest().Method = method
+			seg.Unlock()
+
+			err := invoker(ctx2, method, req, reply, cc, opts...)
+
+			recordContentLength(seg, reply)
+			if err != nil {
+				classifyErrorStatus(seg, err)
+				ct.subsegments.GotConn(nil, err)
+			}
+
+			return err
+		})
+	}
+}
+
+// UnaryServerInterceptor provides gRPC unary server interceptor.
+func UnaryServerInterceptor(ctx context.Context, sn SegmentNamer) grpc.UnaryServerInterceptor {
```

Comment:
AFAIK, the `ctx` here is only used to pass in the value of  `RecorderContextKey`, which should be a `Config`.

Normally, it's a `nil` so the recorder uses the global config.

I can see that for HTTP instrumentation, there's two types of wrapper:
1. [`xray.Handler`](https://pkg.go.dev/github.com/aws/aws-xray-sdk-go/xray#Handler)
2. [`xray.HandlerWithContext`](https://pkg.go.dev/github.com/aws/aws-xray-sdk-go/xray#HandlerWithContext)

Maybe we should break the interceptor into two signatures as well? One without context and another one with context.

Reply:
I think that's a fair call out. We can have two types of wrapper here as well with `ctx` and without `ctx`

Reply:
added `UnaryServerInterceptorWithContext`.